File: library/src/main/java/com/smartmobilefactory/epubreader/display/view/BaseViewPagerAdapter.java
package com.smartmobilefactory.epubreader.display.view;

import android.support.annotation.Nullable;
import android.support.v4.view.PagerAdapter;
import android.util.SparseArray;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

public abstract class BaseViewPagerAdapter extends PagerAdapter {

    private final SparseArray<View> attachedViews = new SparseArray<>();

    @Nullable
    public View getViewIfAttached(int position) {
        return attachedViews.get(position);
    }

    public List<View> getAttachedViews() {
        List<View> views = new ArrayList<>();
        for (int i = 0; i < attachedViews.size(); i++) {
            views.add(attachedViews.valueAt(i));
        }
        return views;
    }

    public abstract View getView(int position, ViewGroup parent);

    public void onItemDestroyed(int position, View view) {

    }

    public Object instantiateItem(ViewGroup collection, int position) {
        View view = getView(position, collection);
        attachedViews.append(position, view);
        collection.addView(view, 0);
        return view;
    }

    @Override
    public void destroyItem(ViewGroup container, int position, Object object) {
        container.removeView((View) object);
        attachedViews.remove(position);
        onItemDestroyed(position, (View) object);
    }

    @Override
    public boolean isViewFromObject(View view, Object object) {
        return view == object;
    }
}

File: library/src/main/java/com/smartmobilefactory/epubreader/display/view/BaseViewPagerAdapter.kt
package com.smartmobilefactory.epubreader.display.view

import android.support.v4.view.PagerAdapter
import android.util.SparseArray
import android.view.View
import android.view.ViewGroup

abstract class BaseViewPagerAdapter : PagerAdapter() {

    private val _attachedViews = SparseArray<View>()

    val attachedViews: List<View>
        get() = (0 until _attachedViews.size()).map { _attachedViews.valueAt(it) }

    fun getViewIfAttached(position: Int): View? {
        return _attachedViews.get(position)
    }

    abstract fun getView(position: Int, parent: ViewGroup): View

    open fun onItemDestroyed(position: Int, view: View) {

    }

    override fun instantiateItem(collection: ViewGroup, position: Int): Any {
        val view = getView(position, collection)
        _attachedViews.append(position, view)
        collection.addView(view, 0)
        return view
    }

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        container.removeView(`object` as View)
        _attachedViews.remove(position)
        onItemDestroyed(position, `object`)
    }

    override fun isViewFromObject(view: View, `object`: Any): Boolean {
        return view === `object`
    }
}